examples/metrics/application_efficiency.py

```diff
@@ -0,0 +1,307 @@
+an application is achieving the best possible performance.
+
+There are multiple ways we can normalize the data to measure relative
+efficiency, and for this tutorial we will consider *application efficiency*.
+
+Application Efficiency
+----------------------
+
+The simplest form of performance efficiency we can work with is *application
+efficiency*, defined below:
+
+**Application Efficiency**
+  The performance of an *application*, measured relative to the best known
+  performance previously demonstrated for solving the same *problem* on the
+  same *platform*.
+
+Working with application efficiency is simple because it does not rely on
+performance models or theoretical hardware limits. Although it can't tell us
+whether an application is performing as well as theoretically possible, it
+shows how an application compares to the state-of-the-art, which is often good
+enough.
+
+Calculating Application Efficiency
+----------------------------------
+
+Let's begin with a very simple example, with a single application, using the
+data below:
+
+.. list-table::
+    :widths: 20 20 20 20 20
+    :header-rows: 1
+
+    * - problem
+      - application
+      - platform
+      - fom
+      - date
+
+    * - Test
+      - MyApp
+      - A
+      - 25.0
+      - 2023
+    * - Test
+      - MyApp
+      - B
+      - 12.5
+      - 2023
+    * - Test
+      - MyApp
+      - C
+      - 25.0
+      - 2023
+    * - Test
+      - MyApp
+      - D
+      - NaN
+      - 2023
+    * - Test
+      - MyApp
+      - E
+      - 5.0
+      - 2023
+
+.. tip::
+    A NaN or 0.0 performance result is interpreted by the P3 Analysis Library
+    to mean that an application run was in some way invalid. We can use this
+    to explicitly represent cases where applications did not compile on
+    specific platforms, did not run to completion, or ran but produced
+    numerical results that failed some sort of verification.
+
+"""
+
+# %%
+# After loading this data into a pandas DataFrame (`df`), we can use the
+# :py:func:`p3.metrics.application_efficiency` function to calculate a table of
+# application efficiencies.
+
+#sphinx_gallery_start_ignore
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+import p3
+
+myapp_data = {
+    "problem": ["Test"] * 5,
+    "application": ["MyApp"] * 5,
+    "platform": ["A", "B", "C", "D", "E"],
+    "fom": [25.0, 12.5, 25.0, np.nan, 5.0],
+    "date": [2023] * 5,
+}
+
+df = pd.DataFrame(myapp_data)
+#sphinx_gallery_end_ignore
+
+effs = p3.metrics.application_efficiency(df)
+print(effs)
+
+# %%
+# These initial results may be a little surprising, because they're all
+# either 1.0 or 0.0. What happened? Since our dataset contains only one result
+# for MyApp on each platform, each non-zero result is the "best known" result
+# for that platform! The only exception is Platform D, which is assigned an
+# efficiency of 0.0 to reflect that it did not run.
```

Comment:
Suggest that this is just clarified slightly

> ... to reflect that it either did not compile, or did not run successfully.

Reply:
Fixed in 7a9729f908acac9bf492d9c26125d226251be4e3.